GVRf/Framework/src/org/gearvrf/GVRHybridObject.java

```diff
@@ -185,6 +196,189 @@ public int getUseCount() {
+
+        private final Map<List<NativeCleanupHandler>, List<NativeCleanupHandler>> //
+        mUniqueCopies = new HashMap<List<NativeCleanupHandler>, List<NativeCleanupHandler>>();
+
+        /**
+         * Typically, descendants have a single (static) list of cleanup
+         * handlers: pass that list to this constructor.
+         * 
+         * @param prefixList
+         *            List of cleanup handler(s)
+         */
+        protected CleanupHandlerListManager(
+                List<NativeCleanupHandler> prefixList) {
+            mPrefixList = prefixList;
+        }
+
+        /**
+         * Descendants that add a cleanup handler list use this method to create
+         * unique concatenations of their list with any of <em>their</em>
+         * descendants' list(s).
+         * 
+         * @param suffix
+         *            Descendant's (static) list
+         * @return A unique concatenation
+         */
+        protected List<NativeCleanupHandler> getUniqueConcatenation(
+                List<NativeCleanupHandler> suffix) {
+            if (suffix == null) {
+                return mPrefixList;
+            }
+
+            List<NativeCleanupHandler> concatenation = mUniqueCopies
+                    .get(suffix);
+            if (concatenation == null) {
+                concatenation = new ArrayList<NativeCleanupHandler>(
+                        mPrefixList.size() + suffix.size());
+                concatenation.addAll(mPrefixList);
+                concatenation.addAll(suffix);
+                mUniqueCopies.put(suffix, concatenation);
+            }
+            return concatenation;
+        }
+    }
+
+    private static class GVRReference extends PhantomReference<GVRHybridObject> {
+
+        // private static final String TAG = Log.tag(GVRReference.class);
+
+        private long mNativePointer;
+        private final List<NativeCleanupHandler> mCleanupHandlers;
+
+        private GVRReference(GVRHybridObject object, long nativePointer,
+                List<NativeCleanupHandler> cleanupHandlers) {
+            super(object, sReferenceQueue);
+
+            mNativePointer = nativePointer;
+            mCleanupHandlers = cleanupHandlers;
+        }
+
+        private void close() {
+            if (mNativePointer != 0) {
+                if (mCleanupHandlers != null) {
+                    for (NativeCleanupHandler handler : mCleanupHandlers) {
+                        handler.nativeCleanup(mNativePointer);
+                    }
+                }
+                NativeHybridObject.delete(mNativePointer);
+            }
+
+            sReferenceSet.remove(this);
+        }
+    }
+
+    private static class GVRFinalizeThread extends Thread {
+
+        // private static final String TAG = Log.tag(GVRFinalizeThread.class);
+
+        private GVRFinalizeThread() {
+            setName("GVRF Finalize Thread");
+            setPriority(MAX_PRIORITY);
+            start();
+        }
```

Comment:
> As a GC thread, imho, I think it is better for you to call setDaemon(true) to make it a daemon thread.

Huh?

    setDaemon (boolean isDaemon)

    Marks this thread as a daemon thread. A daemon thread only runs as long 
    as there are non-daemon threads running. When the last non-daemon 
    thread ends, the runtime will exit. This is not normally relevant to applications
    with a UI.

I don't see how this would be appropriate, here. A GVRf app _is_ an app with a UI, not a background service.

> And I am not quite understanding that why you set this thread with highest priority. Actually in my opinion, the GC thread should run with low priority since we don't want it to slow down or even block the other threads.

This is only going to wake up after GC has freed objects - this typically involves pausing all threads. Since this thread is _mostly_ invoking native `delete`, it seems like a good idea to make this happen as quickly as possible, before (say) low-priority threads allocate more memory.